### ISSUE TITLE
General encoding and hashing for Sway data

### ISSUE THREAD

**Issue by sezna** [FuelLabs/sway#274]
For the next milestone, we need the ability to encode data and hash it with a general enough interface for consumers to adapt to their needs.

From a high level, we need:
1. A memory encoder (pull stack data into heap and make it all contiguous)
1. A hasher for the output of the above type

related: 
#218 
#144 
#217 

blocked by:
#272
#273

**Comment by adlerjohn** [comment]
Superseded by #1511
